examples/vello_editor/src/text.rs

```diff
@@ -326,6 +594,56 @@ impl Editor {
             }
             _ => {}
         }
+
+        if let ImeState::Preedit(ref text_range) = self.ime {
+            if text_range.start != self.selection.anchor().text_range().start {
+                // If the selection anchor is no longer at the same position as the preedit text, the
+                // selection has been moved. Move the preedit to the selection's new anchor position.
+
+                // TODO: we can be smarter here to prevent need of the String allocation
+                let text = self.buffer[text_range.clone()].to_owned();
+                self.buffer.replace_range(text_range.clone(), "");
+
+                if self.selection.anchor().text_range().start > text_range.start {
+                    // shift the selection to the left to account for the preedit text that was
+                    // just removed
+                    let anchor = *self.selection.anchor();
+                    let focus = *self.selection.focus();
+                    let shift = text_range
+                        .len()
+                        .min(anchor.text_range().start - text_range.start);
+                    self.selection = Selection::from_cursors(
+                        Cursor::from_index(&self.layout, anchor.index() - shift, anchor.affinity()),
+                        Cursor::from_index(&self.layout, focus.index() - shift, focus.affinity()),
+                    );
+                }
+
+                let insertion_index = self.selection.insertion_index();
+                self.buffer.insert_str(insertion_index, &text);
+                self.ime = ImeState::Preedit(insertion_index..insertion_index + text.len());
+
+                // TODO: events that caused the preedit to be moved may also have updated the
+                // layout, in that case we're now updating twice.
+                self.update_layout(self.width, 1.0);
+
+                self.selection = self.selection.refresh(&self.layout);
+            }
+        }
+
+        if self.ime != ImeState::Disabled
+            && !matches!(
+                event,
+                WindowEvent::RedrawRequested | WindowEvent::CursorMoved { .. }
```

Comment:
Can  `CursorMoved` change the selection if we're currently dragging?

Reply:
Yes. Currently the selection is also changed if we click somewhere. The editor state remains consistent: the IME preedit is moved to the new anchor position, and the span of the selection itself is mostly a visual thing. The user can't delete, copy, paste, etc., while a preedit is active.

What we should do here is not clear. Moving the preedit text to the new anchor location feels a bit clunky, and that's even with the IME selecting the entire preedit. If only part of the preedit is selected, it'll probably feel even stranger.

Reply:
Ah, I see what you mean. Indeed, I think that won't really work as intended. The comment I wrote in that section: 

```
// TODO: this is a bit overzealous in recalculating the IME cursor area: there are
// cases where it can cheaply be known the area is unchanged. (If the calculated area
// is unchanged from the previous one sent, it is not re-sent to the platform, though).
//
// Ideally this is called only when the layout, selection, or preedit has changed, or
// if the IME has just been enabled.
```

(Clearly it's not only overzealous, it's also slightly buggy.)

Maybe the new `PlainEditor` should make it cheap to know whether the layout changed. I'd prefer calling out to the platform as conservatively as possible, as there's no control over the cost of those calls.